src/hotspot/share/nmt/indexedFreeListAllocator.hpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+
+ */
+
+#include "memory/allocation.hpp"
+#include "memory/arena.hpp"
+#include "utilities/growableArray.hpp"
+#include "runtime/os.hpp"
+
+#ifndef SHARE_NMT_INDEXEDFREELISTALLOCATOR_HPP
+#define SHARE_NMT_INDEXEDFREELISTALLOCATOR_HPP
+
+// A free list, growth only, allocator for a specific type E.
+// The allocator returns 'pointers' of 4-bytes in size, allowing for
+// memory savings if a pointer-heavy self-referential structure is used.
+// It is "indexed" as a reference is base + index * sizeof(E).
+// It never returns any memory to the system.
+template<typename E, MEMFLAGS flag>
+class IndexedFreeListAllocator {
+public:
+  class I {
+    friend IndexedFreeListAllocator<E, flag>;
+    int32_t _idx;
+#ifdef ASSERT
+    IndexedFreeListAllocator<E, flag>* _owner;
+#endif
+
+  public:
+    bool operator !=(I other) {
+      return _idx != other._idx;
+    }
+    bool operator==(I other) {
+      return _idx == other._idx;
+    }
+
+    I(int32_t idx DEBUG_ONLY(COMMA IndexedFreeListAllocator<E COMMA flag>* owner))
+    : _idx(idx) DEBUG_ONLY(COMMA _owner(owner)) {}
+  };
+  static const I nil;
+
+private:
+  // A free list allocator element is either a link to the next free space
+  // Or an actual element.
+  union alignas(E) BackingElement {
+    I link;
+    char e[sizeof(E)];
+
+    BackingElement() {
```

Comment:
Init list?

Reply:
It's on its way, I haven't pushed all of my changes yet :-).